crates/accelerate/src/twirling.rs

```diff
@@ -0,0 +1,472 @@
+                        )),
+                        extra_attrs: inst.extra_attrs.clone(),
+                        #[cfg(feature = "cache_pygates")]
+                        py_op: std::cell::OnceCell::new(),
+                    };
+                    #[cfg(feature = "cache_pygates")]
+                    new_inst.py_op.set(new_inst_obj).unwrap();
+                    out_circ.push(py, new_inst)?;
+                } else {
+                    out_circ.push(py, inst.clone())?;
+                }
+            }
+            _ => {
+                out_circ.push(py, inst.clone())?;
+            }
+        }
+    }
+    if run_pass {
+        let mut dag = DAGCircuit::from_circuit_data(py, out_circ, false)?;
+        optimize_1q_gates_decomposition(
+            py,
+            &mut dag,
+            optimizer_target,
+            optimizer_global_decomposer.cloned(),
+            optimizer_basis_gates.cloned(),
+        )?;
+        dag_to_circuit(py, &dag, false)
+    } else {
+        Ok(out_circ)
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+#[pyfunction]
+#[pyo3(signature=(circ, twirled_gate=None, custom_twirled_gates=None, seed=None, num_twirls=1, run_pass=false, optimizer_target=None, optimizer_global_decomposer=None, optimizer_basis_gates=None))]
+pub(crate) fn twirl_circuit(
+    py: Python,
+    circ: &CircuitData,
+    twirled_gate: Option<Vec<StandardGate>>,
+    custom_twirled_gates: Option<Vec<OperationFromPython>>,
+    seed: Option<u64>,
+    num_twirls: usize,
+    run_pass: bool,
+    optimizer_target: Option<&Target>,
+    optimizer_global_decomposer: Option<HashSet<String>>,
+    optimizer_basis_gates: Option<Vec<String>>,
+) -> PyResult<Vec<CircuitData>> {
+    let mut rng = match seed {
+        Some(seed) => Pcg64Mcg::seed_from_u64(seed),
+        None => Pcg64Mcg::from_entropy(),
+    };
+    let twirling_mask: u8 = match twirled_gate {
+        Some(gates) => {
+            let mut out_mask = 0;
+            for gate in gates {
+                let new_mask = match gate {
+                    StandardGate::CXGate => CX_MASK,
+                    StandardGate::CZGate => CZ_MASK,
+                    StandardGate::ECRGate => ECR_MASK,
+                    StandardGate::ISwapGate => ISWAP_MASK,
+                    _ => {
+                        return Err(QiskitError::new_err(
+                          format!("Provided gate to twirl, {}, is not currently supported you can only use cx, cz, ecr or iswap.", gate.name())
+                        ))
+                    }
+                };
+                out_mask |= new_mask;
+            }
+            out_mask
+        }
+        None => {
+            if custom_twirled_gates.is_none() {
+                15
+            } else {
+                0
+            }
+        }
+    };
+    let custom_gate_twirling_sets: Option<HashMap<String, Vec<([StandardGate; 4], f64)>>> =
+        custom_twirled_gates.map(|gates| {
```

Comment:
Oops, I forgot to add checking on the qubit count and parameters to ensure we only run this with 2q gates that don't take parameters.